freqtrade/rpc/api_server/ws/channel.py

```diff
@@ -0,0 +1,178 @@
+        self._wrapped_ws = self._serializer_cls(self._websocket)
+
+    def __repr__(self):
+        return f"WebSocketChannel({self.channel_id}, {self.remote_addr})"
+
+    @property
+    def remote_addr(self):
+        return self._websocket.remote_addr
+
+    async def send(self, data):
+        """
+        Send data on the wrapped websocket
+        """
+        await self._wrapped_ws.send(data)
+
+    async def recv(self):
+        """
+        Receive data on the wrapped websocket
+        """
+        return await self._wrapped_ws.recv()
+
+    async def ping(self):
+        """
+        Ping the websocket
+        """
+        return await self._websocket.ping()
+
+    async def close(self):
+        """
+        Close the WebSocketChannel
+        """
+
+        self._closed = True
+
+    def is_closed(self) -> bool:
+        """
+        Closed flag
+        """
+        return self._closed
+
+    def set_subscriptions(self, subscriptions: List[str] = []) -> None:
+        """
+        Set which subscriptions this channel is subscribed to
+
+        :param subscriptions: List of subscriptions, List[str]
+        """
+        self._subscriptions = subscriptions
+
+    def subscribed_to(self, message_type: str) -> bool:
+        """
+        Check if this channel is subscribed to the message_type
+
+        :param message_type: The message type to check
+        """
+        return message_type in self._subscriptions
+
+
+class ChannelManager:
+    def __init__(self):
+        self.channels = dict()
+        self._lock = RLock()  # Re-entrant Lock
+
+    async def on_connect(self, websocket: WebSocketType):
+        """
+        Wrap websocket connection into Channel and add to list
+
+        :param websocket: The WebSocket object to attach to the Channel
+        """
+        if isinstance(websocket, FastAPIWebSocket):
+            try:
+                await websocket.accept()
+            except RuntimeError:
+                # The connection was closed before we could accept it
+                return
+
+        ws_channel = WebSocketChannel(websocket)
+
+        with self._lock:
+            self.channels[websocket] = ws_channel
+
+        return ws_channel
+
+    async def on_disconnect(self, websocket: WebSocketType):
+        """
+        Call close on the channel if it's not, and remove from channel list
+
+        :param websocket: The WebSocket objet attached to the Channel
+        """
+        with self._lock:
+            channel = self.channels.get(websocket)
+            if channel:
+                if not channel.is_closed():
+                    await channel.close()
+
+                del self.channels[websocket]
+
+    async def disconnect_all(self):
+        """
+        Disconnect all Channels
+        """
+        with self._lock:
+            for websocket, channel in self.channels.items():
+                if not channel.is_closed():
+                    await channel.close()
```

Comment:
I suspect you might want to wrap the calls to `.close()` in a try/except handler.
If the first "close" call fails - you'd only ever close one channel - leaving all others open (most likely - blocking shutdown)